packages/tunlrs-core/src/tunnel/tunnelconnection.rs

```diff
@@ -46,17 +48,48 @@ impl TunnelConnection {
         }
     }
 
+    fn consume_callback_function(&mut self, func: Option<function_callback>) -> () {
+        match func {
+            Some(x) => x(&self),
+            None => (),
+        }
+    }
+
     pub fn set_on_connect(&mut self) {}
     pub fn set_on_client_read(&mut self) {}
     pub fn set_on_server_request(&mut self) {}
     pub fn set_on_server_response(&mut self) {}
     pub fn set_on_client_write(&mut self) {}
     pub fn set_on_disconnect(&mut self) {}
 
-    pub fn connect() {}
-    pub fn relay_to_server() {}
-    pub fn reply_to_client() {}
-    pub fn disconnect() {}
+    pub fn connect(&mut self) {
+        let callback = self.on_connect.take();
+        self.consume_callback_function(callback);
+        /* do other connect stuff */
+    }
+    pub fn relay_to_server(&mut self) {
+        let mut callback = self.on_client_read.take();
+        self.consume_callback_function(callback);
+        /* do other relay_to_server stuff */
+
+        callback = self.on_server_request.take();
+        self.consume_callback_function(callback);
+        /* do stuff in between request and response */
+        callback = self.on_server_response.take();
+        self.consume_callback_function(callback);
```

Comment:
hmm do a pre req and post req explicit callbacks here, since we have no idea to know whether its being performed after or before the req yet

Reply:
Makes sense, the other network code is supposed to go between these two callbacks, which is why I reassigned it

Btw, we'll have to see that the part between the server request and response is either made blocking or we put the rest of it into .then() blocks